src/main/java/org/spongepowered/api/event/cause/Cause.java

```diff
@@ -48,52 +52,211 @@
+     * cause may mean the {@link Cause} is not originating from any vanilla
+     * interactions, or it may mean the cause is simply not known.
      *
-     * @param parent An optional parent
-     * @param cause The causing object (may be a block, entity, etc.)
-     * @param reason An optional reason
+     * @return True if this cause is empty, false otherwise
      */
-    public Cause(@Nullable Cause parent, Object cause, @Nullable Reason reason) {
-        checkNotNull(cause, "cause");
-        this.parent = Optional.fromNullable(parent);
-        this.cause = cause;
-        this.reason = Optional.fromNullable(reason);
-    }
+    public abstract boolean isEmpty();
 
     /**
-     * Get the parent cause of this cause.
+     * Gets the root {@link Object} of this cause. The root can be anything,
+     * including but not limited to: {@link DamageSource}, {@link Entity},
+     * {@link SpawnCause}, etc.
      *
-     * @return The parent cause
+     * @return The root object cause for this cause
      */
-    public Optional<Cause> getParent() {
-        return this.parent;
-    }
+    public abstract Optional<?> getRoot();
 
     /**
-     * Get the causing object (it may be an {@link Entity}, {@link Location},
-     * etc.).
+     * Gets the first {@link T} object of this {@link Cause}, if available.
      *
-     * @return The cause
+     * @param target The class of the target type
+     * @param <T> The type of object being queried for
+     * @return The first element of the type, if available
      */
-    public Object getCause() {
-        return this.cause;
-    }
+    public abstract <T> Optional<T> getFirst(Class<T> target);
 
     /**
-     * Get the reason.
+     * Gets the last object instance of the {@link Class} of type {@link T}.
      *
-     * @return The reason
+     * @param target The class of the target type
+     * @param <T> The type of object being queried for
+     * @return The last element of the type, if available
+     */
+    public abstract <T> Optional<T> getLast(Class<T> target);
+
+    /**
+     * Gets an {@link ImmutableList} of all objects that are instances of the
+     * given {@link Class} type {@link T}.
+     *
+     * @param target The class of the target type
+     * @param <T> The type of objects to query for
+     * @return An immutable list of the objects queried
+     */
+    public abstract <T> ImmutableList<T> getAllOf(Class<T> target);
+
+    /**
+     * Gets an {@link ImmutableList} of all causes within this {@link Cause}.
+     *
+     * @return An immutable list of all the causes
+     */
+    public abstract ImmutableList<Object> getAllCauses();
+
+    /**
+     * Returns {@code true} if {@code object} is a {@code Cause} instance, and either
+     * the contained references are {@linkplain Object#equals equal} to each other or both
+     * are absent.
+     */
+    @Override
+    public abstract boolean equals(@Nullable Object object);
```

Comment:
Ensures that the methods are properly overridden instead of ignored like most classes/interfaces in the API.